lib/spack/spack/multimethod.py

```diff
@@ -279,6 +279,25 @@ def default_args(**kwargs):
     spack.directives.DirectiveMeta.pop_default_args()
 
 
+every_plat_but_windows = ["cray", "darwin", "freebsd", "linux"]
+
+
+def everywhere_but_windows(directive, *args, **kwargs):
+    """Abstraction around making directives non Windows
+    Invokes directive with arguments given every platform
+    scope but Windows
+
+    Args:
+        directive: function pointer for the
+            directive to be invoked
+        args (list): positional arguments given to directive
+        kwargs (dict): keyword args for directive
+    """
+    for plat in every_plat_but_windows:
+        with when(f"platform={plat}"):
+            directive(*args, **kwargs)
```

Comment:
Do we really need this? I think most packages would just check:
```python
if sys.platform != "win32":
```

Reply:
> Additionally, wouldn't expressing it like this require less refactoring when we do support cross compiling?

I don't think so. It's likely that we need to be able to express negation of a constraint in the syntax directly.

> it either caused errors in some scenarios, or was deemed sub-optimal due to the mix of python and spec logic.

Do you have examples of errors caused by this? For instance I see:

https://github.com/spack/spack/blob/42de252c12435c44f64f2b6b7841f6c06db16569/var/spack/repos/builtin/packages/bzip2/package.py#L46-L47

and that seems fine to me given the current DSL.

Reply:
> we need to be able to express negation of a constraint in the syntax directly.

Working on this, seems like Spack isn't quite ready to support however: https://github.com/spack/spack/pull/41453

> Do you have examples of errors caused by this? For instance I see:

No I don't, and perhaps the "errors" it caused were speculative, as I can't really think of _why_ this would cause errors unless we were cross compiling.
I think the reason we started doing it this way was mostly to future proof against cross compiling. Again, this was a decision made a few years ago, so my memory is fairly hazy.

I do see some value in just having the queries about the current build configuration be scoped to the spec if only just for consistencies sake and to prevent potential conflicts if the behavior of either the disparate platform or spec checks changed.

Reply:
>> `def everywhere_but_windows`

> Do we really need this? I think most packages would just check...

I consider it more idiomatic to examine spec details than to use `sys.platform` (I agree no error should come from using `if sys.platform ==...` though). 

What's wrong with this? I assume it wouldn't affect concretization time because the platform is fixed (since as you say, for now we don't support cross compiling).

I was hoping it would be possible to factor this so a user could write something like:

```
with everywhere_but_windows():
    patch("internal_findHDF5.patch", when="@:8")
```

@alalazo would any such scheme be problematic to you, or do you just dislike the syntax in its current state?

>> Additionally, wouldn't expressing it like this require less refactoring when we do support cross compiling?

> I don't think so. It's likely that we need to be able to express negation of a constraint in the syntax directly.

I agree this probably doesn't make moving to `with when=!platform=windows` easier (grep for `everywhere_but_windows` vs. `sys.platform == ... win32` are both easy) but still like this for the above mentioned reasons.

Reply:
> would any such scheme be problematic to you, or do you just dislike the syntax in its current state?

Can we split the PR, and discuss this function separately? I don't mean to delay any fix on VTK.

> What's wrong with this?

In my opinion:
- This doesn't add much to the current idiom, and adds a clear asymmetry in the API (Why is this only for Windows? Do we need it for other platforms too?)
- It breaks the current idiom for directives, since we need to wrap the directive into another function

Reply:
I'm pulling these changes out into a new MR.

IMO the major benefit here is that we very rarely express "everywhere but darwin" as a concept, but nearly every package port has the concept of "everywhere but Windows" due to the disjoint nature of a non unix platform to unix based platforms. I'd expect some asymmetry when incorporating such a platform. Asymmetry aside, the main benefit here is that we don't have to maintain a list of platforms in several package files. For example, since we added `freebsd` to the list of platforms, there are now several packages that are incorrectly defining variants for the platform because it's not included in my `plat in ["linux", "darwin", "cray"]` loops and no one went back to refactor. Hdf5 is a great example of this, I'd expect freebsd users to have issues using Hdf5 as we're now not filtering the compiler wrappers on freebsd.

Reply:
> It breaks the current idiom for directives, since we need to wrap the directive into another function

I can make this a true directive, I was just iterating quickly to get VTK patched for Windows and didn't want to spend the time at that moment, should not be a point against the concept as a whole